vm/array.go

```diff
@@ -125,31 +151,142 @@ func builtinArrayInstanceMethods() []*BuiltinMethodObject {
 			// a[3] = 20  # => 20
 			// a          # => [10, nil, nil, 20]
 			// a[-2] = 5  # => [10, nil, 5, 20]
+			//
+			// # Double indexing, second argument specify the count of the arguments
+			// a = [1, 2, 3, 4, 5]
+			// a[2, 3] = [1, 2, 3] # <-- Common case
+			// a # => [1, 2, 1, 2, 3]
+			//
+			// a = [1, 2, 3, 4, 5]
+			// a[4, 4] = [1, 2, 3] # <- Exceeded case
+			// a # => [1, 2, 3, 4, 1, 2, 3]
+			//
+			// a = [1, 2, 3, 4, 5]
+			// a[5, 1] = [1, 2, 3] # <-- Edgy case
+			// a # => [1, 2, 3, 4, 5, 1, 2, 3]
+			//
+			// a = [1, 2, 3, 4, 5]
+			// a[8, 123] = [1, 2, 3] # <-- Weak array case
+			// a # => [1, 2, 3, 4, 5, nil, nil, nil, 1, 2, 3]
+			//
+			// a = [1, 2, 3, 4, 5]
+			// a[3, 0] = [1, 2, 3] # <-- Insertion case
+			// a # => [1, 2, 3, 1, 2, 3, 4, 5]
+			//
+			// a = [1, 2, 3, 4, 5]
+			// a[0, 3] = 12345     # <-- Assign non-array value case
+			// a # => [12345, 4, 5]
+			//
+			// a = [1, 2, 3, 4, 5]
+			// a[-3, 2] = [1, 2, 3] # <-- Negative index assign case
+			// a # => [1, 2, 1, 2, 3, 5]
+			//
+			// a = [1, 2, 3, 4, 5]
+			// a[-5, 4] = [9, 8, 7] # <-- Negative index edgy case
+			// a # => [9, 8, 7, 5]
+			//
+			// a = [1, 2, 3, 4, 5]
+			// a[-6, 4] = [9, 8, 7] # <-- Nagative index too small case
+			// # ArgumentError: Index value -6 too small for array. minimum: -5
+			//
+			// # While indexing, it returns nil instead error
+			// a[-6, 4] # => nil
+			//
```

Comment:
Hi, all @goby-lang/maintainers, should we also change this to this kind of behaviour?

```
a = [1, 2, 3, 4, 5];
a[-6, 4] # => ArgumentError: Index value -6 too small for array. minimum: -5
```

Because in Ruby, originally:
```ruby
a = [1, 2, 3, 4, 5];
a[-6, 4] # => nil
```

It returns `nil` instead of errors

Reply:
Should raises error I think

Reply:
This is a very tricky subject.

Ruby consistently never raise an error on Array indexed access (reading); in fact, exceptions are raised only on invalid Array writing.

If it's decided that the out of bound selection on the "left" (negative) side will raise an exception, also the mirrored "right" (positive) side should:

    [0, 1, 2][-3, 2] # => error
    [0, 1, 2][3, 2] #=> error

Ultimately, before taking the decision, it should be clarified very precisely what's the rationale behind Ruby never raising an error on Array access (read), and evaluate if that principle should consisently apply.

Reply:
[Update]
I've pushed a commit in this PR - when indexing array, even if it has only one parameter, if the index value is negatively too small, it will raise the `ArgumentError` no matter what, the same applied to the `ConcurrentArray`:

```
a = [1, 2, 3, 4, 5]

a[-6]
# => ArgumentError: Index value -6 too small for array. minimum: -5
a[-6] = 123
# => ArgumentError: Index value -6 too small for array. minimum: -5
a[-6, 1]
# => ArgumentError: Index value -6 too small for array. minimum: -5
a[-6, 1] = 123
# => ArgumentError: Index value -6 too small for array. minimum: -5

require 'concurrent/array'
ca = Concurrent::Array.new([1, 2, 3, 4, 5])

ca[-6]
# => ArgumentError: Index value -6 too small for array. minimum: -5
ca[-6] = 123
# => ArgumentError: Index value -6 too small for array. minimum: -5
ca[-6, 1]
# => ArgumentError: Index value -6 too small for array. minimum: -5
ca[-6, 1] = 123
# => ArgumentError: Index value -6 too small for array. minimum: -5
```